src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -203,19 +203,37 @@ class ResourceHashtableBase : public STORAGE {
   // the iteration is cancelled.
   template<class ITER>
   void iterate(ITER* iter) const {
+    auto function = [&] (K& k, V& v) {
+      return iter->do_entry(k, v);
+    };
+    iterate(function);
+  }
```

Comment:
Do you still have any usages of this? I'm a bit surprised that the compiler don't complain about not being able to disambiguate calls to this vs the new lambda enabled function.

Reply:
This is still needed as there are many places that use the old style iteration. Also, when the body of the iteration is complex, it might be more readable to do it as a separate class. E.g., here in systemDictionaryShared.cpp:

```
    UnregisteredClassesDuplicationChecker dup_checker;
    _dumptime_table->iterate(&dup_checker);
```

I think there's no ambiguity because `iter->do_entry()` cannot be substituted with the other template, and `function(node->_key, node->_value)` in the other template cannot be substituted here. So by SFINAE the C++ compiler can pick the correct template.

Reply:
[Not a review, just a drive-by comment]
SFINAE operates on arguments and signatures; it doesn't reach into the bodies
of (potential) callees like that. The disambiguation is happening because one
argument (the Closure object) is passed as a pointer, while the other is a
lambda object (not a pointer). In the pointer to closure object case the ITER*
overload is a better match, so is selected. In the lambda case, it's not a
pointer so only the non-pointer overload is applicable.

Reply:
(And SFINAE _can't_ reach into the body like that, since the body might not be available.)

Reply:
@kimbarrett Thanks for the explanation. So we're lucky because we use `ITER*` in the old API. If it were `ITER&` then the C++ compiler would complain about ambiguous overloading.